limitador/src/storage/redis/counters_cache.rs

```diff
@@ -4,26 +4,132 @@ use crate::storage::redis::{
     DEFAULT_MAX_CACHED_COUNTERS, DEFAULT_MAX_TTL_CACHED_COUNTERS_SEC,
     DEFAULT_TTL_RATIO_CACHED_COUNTERS,
 };
+use dashmap::DashMap;
 use moka::sync::Cache;
+use std::collections::HashMap;
+use std::future::Future;
+use std::sync::atomic::{AtomicBool, AtomicI64, Ordering};
 use std::sync::Arc;
 use std::time::{Duration, SystemTime};
+use tokio::select;
+use tokio::sync::Notify;
+use tokio::time::interval;
 
 pub struct CachedCounterValue {
     value: AtomicExpiringValue,
+    initial_value: AtomicI64,
     expiry: AtomicExpiryTime,
 }
 
+pub struct Batcher {
+    updates: DashMap<Counter, Arc<CachedCounterValue>>,
+    notifier: Notify,
+    interval: Duration,
+    priority_flush: AtomicBool,
+}
+
+impl Batcher {
+    fn new(period: Duration) -> Self {
+        Self {
+            updates: Default::default(),
+            notifier: Default::default(),
+            interval: period,
+            priority_flush: AtomicBool::new(false),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.updates.is_empty()
+    }
+
+    pub async fn consume<F, Fut, O>(&self, min: usize, consumer: F) -> O
+    where
+        F: FnOnce(HashMap<Counter, Arc<CachedCounterValue>>) -> Fut,
+        Fut: Future<Output = O>,
+    {
+        let mut interval = interval(self.interval);
+        let mut ready = self.updates.len() >= min;
+        loop {
+            if ready {
+                let mut batch = Vec::with_capacity(min);
+                let mut probably_fake = Vec::with_capacity(min);
+                for entry in &self.updates {
+                    if entry.value().value.ttl() < self.interval {
+                        batch.push(entry.key().clone());
+                        if batch.len() == min {
+                            break;
+                        }
+                    }
+                    if entry.value().expiry.duration() == Duration::from_secs(entry.key().seconds())
```

Comment:
This is bit of crappy hack here…

Reply:
Would it make sense to keep 2 queues within the Batcher, a _priority_ one that will gather the "hot" counters where conditions like this one applies (maybe then tag them with other priority rules) and notify or flush them asap/first ? We might not need to make decisions at "flushing"/consuming and separating into fake batchs. WDYT?

Reply:
I'm working on something... the thing is that "priority" is a time based thing... I'm trying to fold all that logic within `CachedCounterValue`